### PR TITLE
Add Filters to Dependencies Field Resolvers

### PR DESCRIPTION
**Pull request by thomasdax98**
- create factories for `dependencies` and `dependents` resolvers
- add filter args to filter the dependencies and dependents by specific columns
- add paging to dependencies and dependents

### CODE REVIEW

File: packages/api/cms-api/src/dependencies/dependencies.resolver.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        ): Promise<Dependency[]> {
+            return this.dependenciesService.getDependents(node, filter);
+        }
+    }
+
+    return DependentsResolverHost;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function DependenciesAndDependentsResolver<T extends Type<unknown>>(
```
Comment (thomasdax98) [inline packages/api/cms-api/src/dependencies/dependencies.resolver.ts:54]: Is there any way I can avoid this third class that's basically code duplication? For example, by merging the other two classes

File: demo/api/schema.gql [hunk 2]
```diff
@@ -143,7 +143,7 @@ type Page implements DocumentInterface {
   content: PageContentBlockData!
   seo: SeoBlockData!
   createdAt: DateTime!
-  dependencies: [Dependency!]!
+  dependencies(filter: DependenciesFilterInput): [Dependency!]!
```
Comment (johnnyomair) [inline demo/api/schema.gql:146]: For products the filter is only called ProductsFilter, we should use the same naming schema here

**Review by johnnyomair — CHANGES_REQUESTED** [review]
Instead of extending the resolvers, I'd create additional resolvers (it's okay to have multiple resolvers for an entity). You could create two factories:

```ts
DependentsResolverFactory.create(<entity>)
DependenciesResolverFactory.create(<entity>)
```

These factories can then be directly used in the respective modules, for instance

```ts
@Module({
  providers: [
    DependentsResolverFactory.create(News),
    DependenciesResolverFactory.create(News),
  ]
}
class NewsModule {}
```

This would also remove the need for an additional `DependenciesAndDependentsResolver`

File: packages/api/cms-api/src/dependencies/dependencies.service.ts [hunk 3]
```diff
@@ -77,19 +76,71 @@ export class DependenciesService {
...
+    async getDependents(
+        target: AnyEntity<{ id: string }> | { entityName: string; id: string },
+        filter?: {
+            rootGraphqlObjectType?: string;
+            rootEntityName?: string;
+            rootId?: string;
+            rootColumnName?: string;
+        },
+    ): Promise<Dependency[]> {
+        await this.refreshViews();
```
Comment (nsams) [inline packages/api/cms-api/src/dependencies/dependencies.service.ts:88]: If you always refresh views we should not have materialized views at all.

Instead we could:
- refresh after a period of time (eg. 15 minutes or 1 hour)
- refresh in background (like stale while revalidate)
- in the dependencies views offer the user a refesh now button to trigger the refresh
- for other usages (eg. checking dependencies when deleting a dam file) we _need_ up to date data, we need a Api requesting that (eg as argument to the FieldResolver)

**Comment by nsams** [comment]
Inheritance for code reusage must be avoided (generally), much better alternative @johnnyomair suggested.

**Comment by nsams** [comment]
And please explain what the use case for those filters is.

File: packages/api/cms-api/src/dependencies/dependencies.service.ts (hunk 3 above)
Comment (thomasdax98) [reply packages/api/cms-api/src/dependencies/dependencies.service.ts:88]: That's what I implemented here: https://github.com/vivid-planet/comet/pull/1222 

It's the same method, so I didn't make these PRs depending on one another

Comment (nsams) [reply packages/api/cms-api/src/dependencies/dependencies.service.ts:88]: oh, sorry, somehow I completely forgot about that

**Comment by nsams** [comment]
> @nsams I implemented it for usage in https://github.com/vivid-planet/comet/pull/1192. There, I want to only get the DamFile dependencies of a page (and ignore e.g. dependencies on other Pages via links).

:+1:

> I decided to allow more filter options to also cover potential future use cases (e.g. checking if a specific file is used on a specific page)

what about paging and sorting (while we are at it)?

**Comment by thomasdax98** [comment]
> what about paging and sorting (while we are at it)?

I'll add paging. 

Sorting is a difficult topic. Since we don't save any information about the instances in the block index, you can't really sort by anything. Sorting by id doesn't make sense. Sorting by graphql object name would be possible, but since the names are translatable, it would only work correctly for English.

**Comment by thomasdax98** [comment]
@nsams Now I added paging

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**